internal/events/flagged_driver.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/rs/zerolog"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+
+	"github.com/mindersec/minder/internal/engine/engcontext"
+	"github.com/mindersec/minder/internal/events/common"
+	"github.com/mindersec/minder/internal/flags"
+	serverconfig "github.com/mindersec/minder/pkg/config/server"
+)
+
+type flaggedDriver struct {
+	basePub       message.Publisher
+	experimentPub message.Publisher
+	baseSub       message.Subscriber
+	experimentSub message.Subscriber
+
+	flags             openfeature.IClient
+	publishedMessages metric.Int64Counter
+	readMessages      metric.Int64Counter
+}
+
+// Publish implements message.Publisher.  If the message is flagged by the
+// alternate_message_driver experiment, it is published to the experiment
+// driver, otherwise it is published to the base driver.
+func (f *flaggedDriver) Publish(topic string, messages ...*message.Message) error {
+	// Each message has its own context, so they _could_ be in different flag treatments
+	for _, m := range messages {
+		emptyContext := engcontext.EntityContext{}
+		if engcontext.EntityFromContext(m.Context()) == emptyContext {
+			zerolog.Ctx(m.Context()).Warn().Str("stack", string(debug.Stack())).Msg("No entity in context")
+		}
+		if flags.Bool(m.Context(), f.flags, flags.AlternateMessageDriver) {
+			if err := f.experimentPub.Publish(topic, m); err != nil {
+				return err
+			}
+			f.publishedMessages.Add(m.Context(), 1, metric.WithAttributes(attribute.Bool("experiment", true)))
+		} else {
+			if err := f.basePub.Publish(topic, m); err != nil {
+				return err
+			}
+			f.publishedMessages.Add(m.Context(), 1, metric.WithAttributes(attribute.Bool("experiment", false)))
+		}
+	}
+	return nil
+}
+
+// Subscribe implements message.Subscriber.  In this case, it should subscribe
+// to both the base and experiment drivers, because we might get messages for
+// either.
+func (f *flaggedDriver) Subscribe(ctx context.Context, topic string) (<-chan *message.Message, error) {
+	out := make(chan *message.Message)
+	base, err := f.baseSub.Subscribe(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to subscribe to base: %w", err)
+	}
+	experiment, err := f.experimentSub.Subscribe(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to subscribe to experiment: %w", err)
+	}
+	go func() {
+		defer close(out)
+		// Cribbed from https://medium.com/justforfunc/why-are-there-nil-channels-in-go-9877cc0b2308
+		for base != nil || experiment != nil {
+			select {
+			case msg, ok := <-base:
+				if !ok {
+					base = nil
+					continue
+				}
+				out <- msg
+				f.readMessages.Add(ctx, 1, metric.WithAttributes(attribute.Bool("experiment", false)))
+			case msg, ok := <-experiment:
+				if !ok {
+					experiment = nil
+					continue
+				}
+				out <- msg
+				f.readMessages.Add(ctx, 1, metric.WithAttributes(attribute.Bool("experiment", true)))
+			}
+		}
+	}()
```

Comment:
We are missing context cancellation here imo:
```go
case <-ctx.Done():
            return
```

Reply:
It looks like this context is coming from the Router.RunHandlers method, so we should only get a cancellation when the router is shutting down.  (That should also cause all the subscribers to be closed, but it's safe to add this here as well.)

Reply:
(I added the shutdown anyway)